api/v1alpha4/types.go

```diff
@@ -507,3 +508,83 @@ type AddressRecord struct {
+// This eventually gets converted to CloudProviderRateLimitConfig that cloud-provider-azure expects.
+// See: https://github.com/kubernetes-sigs/cloud-provider-azure/blob/d585c2031925b39c925624302f22f8856e29e352/pkg/provider/azure_ratelimit.go#L25
+// We cannot use CloudProviderRateLimitConfig directly because floating point values are not supported in controller-tools.
+// See: https://github.com/kubernetes-sigs/controller-tools/issues/245
+type RateLimitSpec struct {
+	Name   string          `json:"name,omitempty"`
+	Config RateLimitConfig `json:"config,omitempty"`
+}
+
+// RateLimitConfig indicates the rate limit config options.
+type RateLimitConfig struct {
+	CloudProviderRateLimit            bool               `json:"cloudProviderRateLimit,omitempty"`
+	CloudProviderRateLimitQPS         *resource.Quantity `json:"cloudProviderRateLimitQPS,omitempty"`
+	CloudProviderRateLimitBucket      int                `json:"cloudProviderRateLimitBucket,omitempty"`
+	CloudProviderRateLimitQPSWrite    *resource.Quantity `json:"cloudProviderRateLimitQPSWrite,omitempty"`
+	CloudProviderRateLimitBucketWrite int                `json:"cloudProviderRateLimitBucketWrite,omitempty"`
+}
+
+const (
+	// DefaultRateLimit ...
+	DefaultRateLimit = "defaultRateLimit"
+	// RouteRateLimit ...
+	RouteRateLimit = "routeRateLimit"
+	// SubnetsRateLimit ...
+	SubnetsRateLimit = "subnetsRateLimit"
+	// InterfaceRateLimit ...
+	InterfaceRateLimit = "interfaceRateLimit"
+	// RouteTableRateLimit ...
+	RouteTableRateLimit = "routeTableRateLimit"
+	// LoadBalancerRateLimit ...
+	LoadBalancerRateLimit = "loadBalancerRateLimit"
+	// PublicIPAddressRateLimit ...
+	PublicIPAddressRateLimit = "publicIPAddressRateLimit"
+	// SecurityGroupRateLimit ...
+	SecurityGroupRateLimit = "securityGroupRateLimit"
+	// VirtualMachineRateLimit ...
+	VirtualMachineRateLimit = "virtualMachineRateLimit"
+	// StorageAccountRateLimit ...
+	StorageAccountRateLimit = "storageAccountRateLimit"
+	// DiskRateLimit ...
+	DiskRateLimit = "diskRateLimit"
+	// SnapshotRateLimit ...
+	SnapshotRateLimit = "snapshotRateLimit"
+	// VirtualMachineScaleSetRateLimit ...
+	VirtualMachineScaleSetRateLimit = "virtualMachineScaleSetRateLimit"
+	// VirtualMachineSizesRateLimit ...
+	VirtualMachineSizesRateLimit = "virtualMachineSizesRateLimit"
+	// AvailabilitySetRateLimit ...
+	AvailabilitySetRateLimit = "availabilitySetRateLimit"
+)
+
+// allowedToOverrideRateLimits is a collection of rate limit spec names that are allowed to be overridden.
+var allowedToOverrideRateLimits = map[string]bool{
+	DefaultRateLimit: true, RouteRateLimit: true, SubnetsRateLimit: true, InterfaceRateLimit: true, RouteTableRateLimit: true,
+	LoadBalancerRateLimit: true, PublicIPAddressRateLimit: true, SecurityGroupRateLimit: true, VirtualMachineRateLimit: true,
+	StorageAccountRateLimit: true, DiskRateLimit: true, SnapshotRateLimit: true, VirtualMachineScaleSetRateLimit: true,
+	VirtualMachineSizesRateLimit: true, AvailabilitySetRateLimit: true,
+}
```

Comment:
would it be easier to use kubebuilder enum validation (similar to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/api/v1alpha4/types.go#L324) 

makes it a littler easier to read